internal/icinga2/client.go

```diff
@@ -0,0 +1,385 @@
+			client.Logger.Warnw("Catching up the API was interrupted", zap.Duration("duration", time.Since(startTime)))
+		} else {
+			client.Logger.Errorw("Catching up the API failed", zap.Error(err), zap.Duration("duration", time.Since(startTime)))
+
+			select {
+			case <-ctx.Done():
+			case catchupEventCh <- &catchupEventMsg{error: err}:
+			}
+		}
+
+		cancel()
+		close(catchupEventCh)
+	}()
+
+	return catchupEventCh, cancel
+}
+
+// worker is the Client's main background worker, taking care of event.Event dispatching and mode switching.
+//
+// When the Client is in the catch-up-phase, requested by catchupPhaseRequest, events from the Event Stream API will
+// be cached until the catch-up-phase has finished, while replayed events will be delivered directly.
+//
+// Communication takes place over the eventDispatcherEventStream and catchupPhaseRequest channels.
+func (client *Client) worker() {
+	var (
+		// catchupEventCh either emits events generated during the catch-up-phase from catch-up-workers or one final
+		// error if something went wrong. It will be closed when catching up is done, which indicates the select below
+		// to switch phases. When this variable is nil, this Client is in the normal operating phase.
+		catchupEventCh chan *catchupEventMsg
+		// catchupCancel cancels, if not nil, all running catch-up-workers, e.g., when restarting catching-up.
+		catchupCancel context.CancelFunc
+
+		// catchupBuffer holds Event Stream events to be replayed after the catch-up-phase has finished.
+		catchupBuffer = make([]*event.Event, 0)
+		// catchupCache maps event.Events.Name to API time to skip replaying outdated events.
+		catchupCache = make(map[string]time.Time)
+
+		// catchupErr might hold an error received from catchupEventCh, indicating another catch-up-phase run.
+		catchupErr error
+	)
+
+	// catchupReset resets all catchup variables to their initial empty state.
+	catchupReset := func() {
+		catchupEventCh, catchupCancel = nil, nil
+		catchupBuffer = make([]*event.Event, 0)
+		catchupCache = make(map[string]time.Time)
+		catchupErr = nil
+	}
+
+	// catchupCacheUpdate updates the catchupCache if this eventMsg seems to be the latest of its kind.
+	catchupCacheUpdate := func(ev *eventMsg) {
+		ts, ok := catchupCache[ev.event.Name]
+		if !ok || ev.apiTime.After(ts) {
+			catchupCache[ev.event.Name] = ev.apiTime
+		}
+	}
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing down main worker as context is finished", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.catchupPhaseRequest:
+			if catchupEventCh != nil {
+				client.Logger.Warn("Switching to catch-up-phase was requested while already catching up, restarting phase")
+
+				// Drain the old catch-up-phase producer channel until it is closed as its context will be canceled.
+				go func(catchupEventCh chan *catchupEventMsg) {
+					for _, ok := <-catchupEventCh; ok; {
+					}
+				}(catchupEventCh)
```

Comment:
Wait, am I missing something completely or does that loop try to receive one element and then goes into an endless loop depending on whether that worked?

`for range catchupEventCh {}` should actually drain the channel, but is it actually necessary? Is there anything that's sending into the channel that has to be unblocked by this (i.e. doesn't do so in a select together with the context cancel channel)? If not, shouldn't just cancelling the context suffice?

Reply:
You are totally right. If I have to guess how this happened, I would say the `_, ok := <- catchupEventCh` part was prior within a `for`/`select` and messy refactoring resulted in this.

But even after reading both the git log as well the code, I don't see any real reason to drain the channel, as every writing channel access is being protected by a `<-ctx.Done():` check.